docs/root/version_history/current.rst

```diff
@@ -28,6 +28,7 @@ Minor Behavior Changes
 * http: lazy disable downstream connection reading in the HTTP/1 codec to reduce unnecessary system calls. This behavioral change can be temporarily reverted by setting runtime guard ``envoy.reloadable_features.http1_lazy_read_disable`` to false.
 * http: now the max concurrent streams of http2 connection can not only be adjusted down according to the SETTINGS frame but also can be adjusted up, of course, it can not exceed the configured upper bounds. This fix is guarded by ``envoy.reloadable_features.http2_allow_capacity_increase_by_settings``.
 * http: when writing custom filters, `injectEncodedDataToFilterChain` and `injectDecodedDataToFilterChain` now trigger sending of headers if they were not yet sent due to `StopIteration`. Previously, calling one of the inject functions in that state would trigger an assertion. See issue #19891 for more details.
+* listener: convert the IPv4 mapped IPv6 address as IPv4 address for new connection's local address when the listener is listening on IPv4-mapped IPv6 address with ``ipv4_compat`` flag. This behavior matches the case of the listener is listening on IPv6 any address with ``ipv4_compat`` flag. The runtime flag ``envoy.reloadable_features.convert_ipv4_mapped_address_to_ipv4_address`` can be used to revert this behavior.
```

Comment:
I'm only marginally familiar with the compat paths, but I'm a little confused here. How are these cases different? Is the difference that the former is when someone is using a compat address directly, whereas the latter is when someone connects through the compat address but the address itself is not compat (I guess it would have to be any address)?

/wait-any

Reply:
It has nothing to do with the client. The OS(at least linux) always normalizes the address.

It's all about Envoy listener. The behavior is different among LDS config is "any" and a "::ffff::A.B.C.D" where A.B.C.D is a ipv4 address.

I think the root cause is that Envoy doesn't always read the local address using socket() api.
Envoy set local address from ListenerConfig occasionally which is good that it saves a syscall.

Correct me if I am wrong @soulxu

Reply:
> It has nothing to do with the client. The OS(at least linux) always normalizes the address.

Yes, it is nothing to do with the client, but the OS didn't normalize the address. Actually, it is normalized by envoy.

For example, when the ipv4 client sends a request to "::", the listening socket will get a new connection with IPv4-mapped address (OS convert that ipv4 address to ipv4-mapped address). Then envoy convert it back to ipv4 address at
https://github.com/envoyproxy/envoy/blob/80d3c1750a18915f37fa65ca9d1bf3c26dd87d0b/source/common/network/address_impl.cc#L71-L82

> 
> It's all about Envoy listener. The behavior is different among LDS config is "any" and a "::ffff::A.B.C.D" where A.B.C.D is a ipv4 address.
> 
> I think the root cause is that Envoy doesn't always read the local address using socket() api. Envoy set local address from ListenerConfig occasionally which is good that it saves a syscall.
> 
> Correct me if I am wrong @soulxu

Yes, that is exactly what I understand also.

But I begin to doubt the use case of listening on a IPv4-mapped address. Since I found a listener on "192.168.0.1" can get the request from "::FFFF:192.168.0.1", then I don't know what is meaning for listening on "::FFFF:192.168.0.1"

`ipv4_compat` flag is mapping to the OS `IPV6_V6ONLY` flag.  I think the two usecases for this flag are:
1. A IPv6 socket can connect to a IPv4 socket by IPv4-mapped address. When IPV6_V6ONLY as false.
2. For compatibility with IPv4 only applications, I can set IPV6_V6ONLY as true, then I can create a socket on the same port with IPv4-only app one in IPv6-only app. Then IPv6 only app can help to support IPv6 requests. But this seems only meaningful on any address case. The normal IPv6 address won't get the same port space as the IPv4 address. The IPv4-mapped address can't bind with IPV6_V6ONLY as true, the OS reject directly.

We use IPV6_V6ONLY as false and listen on "::" address to accept connections both from IPv4 and IPv6 requests, this seems isn't the major use case of `IPV6_V6ONLY`, but it is totally fine, it perfect for us to use one listener on two IP versions.

Reply:
@lambdai I'm gonna to close this PR, since I still can't ensure listening on IPv4-mapped address is valid usecase, if it isn't valid usecase, it isn't worth to fix that.

Reply:
thanks for the feedback! let me close it, if can't find valid usecase, I don't want to make the code complex also.